// graphcool-tmp-ui/lib/Modal.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var Modal = require("react-modal");
var modalStyle = {
    overlay: {
        zIndex: 20,
        backgroundColor: 'rgba(15,32,46,.9)',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center'
    },
    content: {
        position: 'relative',
        width: 976,
        height: 'auto',
        top: 'initial',
        left: 'initial',
        right: 'initial',
        bottom: 'initial',
        borderRadius: 2,
        padding: 0,
        border: 'none',
        background: 'none',
        boxShadow: '0 1px 7px rgba(0,0,0,.2)'
    }
};
var ModalComponent = function ModalComponent(_ref) {
    var isOpen = _ref.isOpen,
        onRequestClose = _ref.onRequestClose,
        contentLabel = _ref.contentLabel,
        children = _ref.children;

    return React.createElement(
        Modal,
        { isOpen: isOpen, onRequestClose: onRequestClose, contentLabel: contentLabel, style: modalStyle },
        children
    );
};
exports.default = ModalComponent;
