// graphcool-tmp-ui/lib/Tooltip.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _style = require("styled-jsx/style");

var _style2 = _interopRequireDefault(_style);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var ReactDOM = require("react-dom");
var cx = require("classnames");

var Tooltip = function (_React$Component) {
  _inherits(Tooltip, _React$Component);

  function Tooltip() {
    _classCallCheck(this, Tooltip);

    return _possibleConstructorReturn(this, (Tooltip.__proto__ || Object.getPrototypeOf(Tooltip)).apply(this, arguments));
  }

  _createClass(Tooltip, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      document.addEventListener('click', this.handleClickOutside.bind(this), true);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      document.removeEventListener('click', this.handleClickOutside.bind(this), true);
    }
  }, {
    key: "handleClickOutside",
    value: function handleClickOutside(event) {
      if (!this.props.open) {
        return;
      }
      var domNode = ReactDOM.findDOMNode(this);
      if (!domNode || !domNode.contains(event.target)) {
        event.stopPropagation();
        this.props.onClose(event);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _props = this.props,
          open = _props.open,
          children = _props.children,
          onClick = _props.onClick,
          anchorOrigin = _props.anchorOrigin;

      return React.createElement(
        "div",
        { className: cx('tooltip-container', {
            visible: open,
            'anchor-top': anchorOrigin.vertical === 'top',
            'anchor-bottom': anchorOrigin.vertical === 'bottom',
            'anchor-left': anchorOrigin.horizontal === 'left',
            'anchor-right': anchorOrigin.horizontal === 'right',
            'anchor-center': anchorOrigin.horizontal === 'center'
          }), "data-jsx": 3207991532
        },
        React.createElement(_style2.default, {
          styleId: 3207991532,
          css: ".tooltip-container[data-jsx=\"3207991532\"] {-webkit-transform: translateX(-50%);transform: translateX(-50%);z-index: 9999;transition: opacity ease-out 0.2s;visibility: hidden;opacity: 0;}.tooltip-container.anchor-top[data-jsx=\"3207991532\"] {margin-bottom: 16px}.tooltip-container.anchor-top[data-jsx=\"3207991532\"][data-jsx=\"3207991532\"] .big-triangle[data-jsx=\"3207991532\"] {bottom: -10px;}.tooltip-container.anchor-bottom[data-jsx=\"3207991532\"] {margin-top: 16px}.tooltip-container.anchor-bottom[data-jsx=\"3207991532\"][data-jsx=\"3207991532\"] .big-triangle[data-jsx=\"3207991532\"] {top: -10px;border-width: 0 10px 10px 10px;border-color: #f3f4f4 transparent #f3f4f4 transparent;}.tooltip-container.anchor-left[data-jsx=\"3207991532\"] {left: 0;-webkit-transform: none;transform: none}.tooltip-container.anchor-left[data-jsx=\"3207991532\"][data-jsx=\"3207991532\"] .big-triangle[data-jsx=\"3207991532\"] {left: 25px;}.tooltip-container.anchor-right[data-jsx=\"3207991532\"] {right: 0;-webkit-transform: none;transform: none}.tooltip-container.anchor-right[data-jsx=\"3207991532\"][data-jsx=\"3207991532\"] .big-triangle[data-jsx=\"3207991532\"] {right: 25px;}.tooltip-container.anchor-center[data-jsx=\"3207991532\"][data-jsx=\"3207991532\"] .big-triangle[data-jsx=\"3207991532\"] {left: calc(50% - 10px);}.tooltip-container.visible[data-jsx=\"3207991532\"] {visibility: visible;opacity: 1;}.tooltip-container[data-jsx=\"3207991532\"] .tooltip-content[data-jsx=\"3207991532\"] {background-color: #f3f4f4;box-shadow: 0 1px 6px 0 rgba(0, 0, 0, .15);}.big-triangle[data-jsx=\"3207991532\"] {width: 0;height: 0;border-style: solid;border-width: 10px 10px 0 10px;border-color: #f3f4f4 transparent transparent transparent;}.absolute,\n.tooltip-container[data-jsx=\"3207991532\"],\n.big-triangle[data-jsx=\"3207991532\"] {position: absolute;}.tl,\n.tooltip-container[data-jsx=\"3207991532\"] {text-align: left;}.bottom100,\n.tooltip-container.anchor-top[data-jsx=\"3207991532\"] {bottom: 100%;}.top100,\n.tooltip-container.anchor-bottom[data-jsx=\"3207991532\"] {top: 100%;}.left50,\n.tooltip-container.anchor-left[data-jsx=\"3207991532\"],\n.tooltip-container.anchor-center[data-jsx=\"3207991532\"] {left: 50%;}.right50,\n.tooltip-container.anchor-right[data-jsx=\"3207991532\"] {right: 50%;}.left50,\n.tooltip-container.anchor-left[data-jsx=\"3207991532\"],\n.tooltip-container.anchor-center[data-jsx=\"3207991532\"] {left: 50%;}.nowrap,\n.tooltip-container[data-jsx=\"3207991532\"] .tooltip-content[data-jsx=\"3207991532\"] {white-space: nowrap;}.br2,\n.tooltip-container[data-jsx=\"3207991532\"] .tooltip-content[data-jsx=\"3207991532\"] {border-radius: 2px;}.black50,\n.tooltip-container[data-jsx=\"3207991532\"] .tooltip-content[data-jsx=\"3207991532\"] {color: rgba(0,0,0,.5);}.ph16,\n.tooltip-container[data-jsx=\"3207991532\"] .tooltip-content[data-jsx=\"3207991532\"] {padding-left: 16px;padding-right: 16px;}.pv12,\n.tooltip-container[data-jsx=\"3207991532\"] .tooltip-content[data-jsx=\"3207991532\"] {padding-top: 12px;padding-bottom: 12px;}.flex,\n.tooltip-container[data-jsx=\"3207991532\"] .tooltip-content[data-jsx=\"3207991532\"] {display: -ms-flexbox;display: flex;}.itemsCenter,\n.tooltip-container[data-jsx=\"3207991532\"] .tooltip-content[data-jsx=\"3207991532\"] {-webkit-box-align: center;-ms-flex-align: center;-ms-grid-row-align: center;align-items: center;}.absolute,\n.tooltip-container[data-jsx=\"3207991532\"],\n.big-triangle[data-jsx=\"3207991532\"] {position: absolute;}"
        }),
        React.createElement(
          "div",
          { className: "tooltip-content", onClick: onClick, "data-jsx": 3207991532
          },
          React.createElement("div", { className: "big-triangle", "data-jsx": 3207991532
          }),
          children
        )
      );
    }
  }]);

  return Tooltip;
}(React.Component);

Tooltip.defaultProps = {
  anchorOrigin: {
    vertical: 'top',
    horizontal: 'center'
  }
};
exports.default = Tooltip;
